core-bundle/src/EventListener/DataContainer/Undo/LabelListener.php

```diff
@@ -0,0 +1,229 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer\Undo;
+
+use Contao\ArrayUtil;
+use Contao\Config;
+use Contao\Controller;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\DataContainer;
+use Contao\Date;
+use Contao\StringUtil;
+use Contao\System;
+use Contao\UserModel;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver\Exception as DriverException;
+use Doctrine\DBAL\Exception;
+
+/**
+ * @Callback(target="list.label.label", table="tl_undo")
+ *
+ * @internal
+ */
+class LabelListener
+{
+    use UndoListenerTrait;
+
+    private Connection $connection;
+    private ContaoFramework $framework;
+
+    public function __construct(ContaoFramework $framework, Connection $connection)
+    {
+        $this->framework = $framework;
+        $this->connection = $connection;
+    }
+
+    public function __invoke(array $row, string $label, DataContainer $dc): string
+    {
+        $this->framework->initialize();
+
+        $table = $row['fromTable'];
+        $originalRow = StringUtil::deserialize($row['data'])[$table][0];
+
+        /** @var Controller $controller */
+        $controller = $this->framework->getAdapter(Controller::class);
+        $controller->loadDataContainer($table);
+        $controller->loadLanguageFile($table);
+
+        $dataContainer = $this->framework->getAdapter(DataContainer::class)->getDriverForTable($table);
+        $fromTableDc = new $dataContainer($table);
+
+        $header = $this->renderHeader($table, $row, $originalRow);
+        $newLabel = $this->renderLabel($originalRow, $fromTableDc);
+
+        if ($GLOBALS['TL_DCA'][$table]['list']['label']['showColumns'] ?? false) {
+            $newLabel = $this->renderColumns($newLabel);
+        }
+
+        return $header.'<div class="tl_undo_preview">'.$newLabel.'</div>';
+    }
+
+    private function renderHeader(string $table, array $row, array $originalRow): string
```

Comment:
I don't like rendering HTML in PHP. I know it's done in the DC files but we should IMHO not introduce more code with this pattern.

All of this can be done in templates instead. There you already have all the tools for date formatting, translations and encoding (which is missing here as well). Plus: everyone can easily adjust it to their needs.